[Cooper] 미션4 : 모든 기물 배치하기

pbg0205 opened the pull request:
안녕하세요
2021 코드스쿼드 마스터즈 백엔드 과정의 Cooper 입니다.
미션4 코드 리뷰 요청 드립니다.
잘 부탁드립니다!

### 주요 변경 사항
1. Type  Enum을 추가하였습니다.
체스의 기물 특성상 기물의 갯수가 정해져 있어 Enum으로 관리하고 내부 representation을 상수 멤버로 추가하는 것이 적합하다고 생각했습니다.
그래서 이전 미션3에서 Piece 생성자를 문자열로 입력받아 처리하던 방식을 Color와 Type을 매개인자로 받아 인스턴스를 생성하는 방식으로 변경 완료하였습니다.

2. Piece class 내 팩토리 메서드를 선언하였습니다.
 팩토리 메서드를 선언하여 각각 기물에 대한 팩토리 메서드를 선언하였습니다. 메서드를 작성할 때 매개인자 유무 방식에 고민하며 메서드를 작성하였습니다. 팩토리 메서드의 장점은 인스턴스의 캡슐화와 이름을 부여해서 의미가 명확해지는 것으로 알고있습니다. 팩토리 메서드의 장점을 고려해보았을 때, 매개인자를 없는 방식이 캡슐화에 더 이점이 있다고 생각했습니다. 매개인자를 받을 경우, 매개인자에 따라서 반환하는 객체의 종류가 변하기 때문에 캡슐화의 이점이 떨어진다고 생각하였습니다. 그리고 각각의 메서드를 선언한 방법이 각 인스턴스를 생성하는데 의미를 명확하게 전달한다고 생각되어 매개인자 없는 메서드를 선언하였습니다. 우선적으로 위의 내용에 대해서 생각하여 구현했지만 어떤 방법이 맞는 방법이 더욱 좋은 방법인지 여쭙고 싶습니다!

3. Board 전체 기물 상태 반환
 Board 전체 기물 상태를 출력하기 이전에 기물을 관리하는 자료구조에 대해 고민해보았습니다. Pawn와 그 외 기물을 별도로 관리하는 방식과 모든 기물을 관리하는 방식에 대해 고민 해보았지만 현재 단계에서는 리스트를 하나로 관리해도 무방하다고 판단하여 리스트를 하나로 관리하였습니다.

pbg0205 committed:
feat : \n 제거를 위한 StringUtils 클래스 추가

- line.seperator를 사용해 OS에 적합한 줄바꿈을 처리할 수 있도록
구현했습니다.

- 내부에 StringBuilder를 static 변수로 선언하여 해당 객체에 board 상태
를 저장함과 동시에 줄바꿈 연산을 할 수 있도록 구현하였습니다.

pbg0205 committed:
style : Pawn 클래스를 Piece 클래스로 재명명

- Pawn 클래스를 Piece 클래스로 재명명 하였습니다.

- 이전에 Pawn 객체를 관리하던 PawnGroup 클래스를 PieceGroup으로
재명명 하였습니다.

pbg0205 committed:
style : line seperator 접근 제어자 변경

- 멤버변수 NEWLINE과 StringBuider의 접근제어자를 private로
변경하였습니다.

pbg0205 committed:
feat : Type Enum 생성

- 기물의 representation을 별도로 관리하기 위해 Type Enum을
 생성하였습니다.

pbg0205 committed:
refactor : Color 상수 멤버 representaion 제거

- 여러 기물이 생성되어 표시해주어야하기 때문에 별도로 Type Enum을
관리하였습니다.

- Type Enum에서 Representation을 관리하기 때문에 Color에 존재하는
representation을 제거하였습니다.

pbg0205 committed:
refactor : Piece constuctor 매개인자 변경

- 해당 미션부터 represetation을 별도로 관리하기 때문에 기존에 Color만
매개인자로 받던 생성자를 Color와 Type을 매개인자로 받는 생성자로 변경
하였습니다.

- 매개인자를 Color enum으로 받아 멤버 변수를 초기화하기 때문에 문자열을
매개인자로 받던 생성자의 예외 테스트 코드를 제거하였습니다.

- 기본 생성자를 제거하는 요구사항에 맞춰 기본 생성자 제거 및 관련
테스트 코드를 제거하였습니다.

pbg0205 committed:
fix : KINGHT 상수 representation 수정

- KINGHT 상수 멤버명 k을 n으로 변경.

pbg0205 committed:
style : pawnListMap 멤버 변수명:  pieceListMap변경

pbg0205 committed:
refactor : Color enum 상수 멤버 제거

- Piece의 생성자의 매개인자를 Color enum 상수를 사용하기 때문에 불필요한
상수 멤버의 문자열을 제거하였습니다.

pbg0205 committed:
feat : 기물 생성 팩토리 메서드 생성

- 팩토리 메서드를 선언해서 각각의 기물의 인스턴스를 생성할 수 있도록
구현하였습니다.

- 기존 생성자로 객체를 생성하던 방식을 모두 팩토리 메서드를 사용하는
방식으로 변경했습니다.

pbg0205 committed:
fix : 팩터리 내부 생성자 아규먼트 재설정

- createBlackKnight 내부 생성자 아규먼트를 Type.KNIGHT로 수정하였습니다.

- 함수명 createWhiterKnight의 오탈자 수정하였습니다.

pbg0205 committed:
test : 모든 기물 생성 확인 테스트

- 모든 기물이 정확한 생성 여부와 내부 값을 확인하기 위해 테스트 코드를
작성하였습니다.

- 기존의 String과 char로 입력받아 유효성 검사하는 방식보보다 enum
멤버에 한해서 정확한 값을 지정한 범위 내의 파라미터로 입력받는  것이
효율적이라고 생각하여 enum 상수를 파라미터로 변경하였습니다.

pbg0205 committed:
feat : PieceGroup클래스 상수변경 및  size 메서드

- PieceGroup 내의 전체 요소의 갯수를 반환하는 size 메서드를
추가 하였습니다.

- 흰색와 검은색의 List에 최대 요소 갯수를 나타내는 상수를 추가
하였습니다.

pbg0205 committed:
feat : Board 기물 초기화+상태 반환 메서드 추가

- 한 색상에 두 리스트로 선언하는 방법도 고려해보았습니다. 하지만
폰과 나머지 기물을 두 리스트로 관리하는 것의 큰 장점을 생각하지 못했습
니다. 그래서 우선적으로 한 리스트로 선언하여 관리하는 방식으로 접근하려
고 계획하고 있습니다.

- PieceGroup의 전체 요소들을 전달하는 pieceCount 메서드를 추가
하였습니다.

- 기존의 전체 Board의 상태를 반환하는 메서드의 로직을 StringUtils의
appendNewLine 메서드를 통해 전체적으로 상태를 입력해서 반환할 수 있도록
구현하였습니다.

pbg0205 committed:
test : BoardTest - showBoard 확인 테스트 추가

- Board 테스트를 확인하는 클래스이므로 @BeforeEach를 선언해서 각각의
테스트 이전에 Board 인스턴스를 생성하도록 변경하였습니다.

- Board 내 전체 기물의 요소의 갯수를 확인하는 테스트 코드를 create 메서드
내에 추가하였습니다.

- Board 내 전체 기물 상태를 확인할 수 있도록 대조군을 StringBuilder로
생성하고 StringUtils 내에 실제 Board 상태에 대한 정보를 추가 및 비교하
도록 로직을 구현하였습니다.

tmdgusya commented on src/main/java/cooper/chess/piece/PieceGroup.java:
```diff
@@ -4,8 +4,7 @@
 import java.util.Map;
 
 public class PieceGroup {
```
개인적인 생각인데, PieceGroup 이 현재 1급 컬렉션을 만들기 위해 사용된다는 생각이 드는데
getPawnList 로 외부에서 List 를 받은 뒤 해당 List에 Piece 를 추가하는 일들은 Board 에서 직접 일어나고 있는데,
이 경우 add() 에서 유효성 검사의 로직이 필요할 경우 Board 에 유효성 추가 검사 로직이 작성될 것으로 보이는데,
차라리 PieceGroup 에 add() 를 만들고, PieceGroup 내에 검사로직을 만드는 것이 어떨까요?
그렇게 되면 PieceGroup 의 add() 를 쓰는 모든 곳에서 유효성 검사 로직을 작성할 필요가 없어보이는데요.

지금 현재 코드상으로는 외부에서 객체를 통한게 아닌 해당 Collection Element에 변경이 가능하다고 생각됩니다.

tmdgusya commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -17,52 +17,70 @@ public Board() {
         }
 
-        return pieceList;
+        pieceList.add(Piece.createBlackRook());
+        pieceList.add(Piece.createBlackKnight());
+        pieceList.add(Piece.createBlackBishop());
+        pieceList.add(Piece.createBlackQueen());
+        pieceList.add(Piece.createBlackKing());
+        pieceList.add(Piece.createBlackBishop());
+        pieceList.add(Piece.createBlackKnight());
+        pieceList.add(Piece.createBlackRook());
```
제가 설명드린 부분입니다.

tmdgusya commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -73,33 +91,28 @@ public Piece findPawn(int index, Color color) {
-        while(index < PAWN_MAX_SIZE) {
-            appendNewLine(getEachRowStatus(index++));
-        }
+    public String showBoard() {
```
궁금한것이 지금 Main.java 에서 해당 함수를 반복문으로 여러번 호출하고 있는데,
이렇게 StringUtils 에서 static 하게 StringBuilder 필드를 가지고 있는 상태에서 showBoard() 를 반복하게 되면
메세지가 중첩되서 체스판이 반복문 돌면서 showBoard() 를 수행하면서 추가될것이라고 생각되는데,
혹시 이부분 테스트 해보셨나요?

간단히 얘기하면 체스판 1개만 나와야되는데 Main.class 의 로직이 반복될때마다 2개 3개 4개... 이런식으로 나올것 같아보입니다.

tmdgusya commented on src/test/java/cooper/chess/BoardTest.java:
```diff
@@ -1,47 +1,53 @@
-                    assertEquals(white, board.findPawn(0, Color.WHITE));
-                    assertEquals(black, board.findPawn(0, Color.BLACK));
-                }
+        sb.append("RNBQKBNR").append("\n");
```
"\n" 여기 부분 차라리 StringUtils 의 NEWLINE 을 public 으로 열고 NEWLINE 으로 적어보는건 어떨까요?

그냥 개인적인 견해입니다

tmdgusya commented on src/test/java/cooper/chess/BoardTest.java:
```diff
@@ -1,47 +1,53 @@
+
+        assertEquals(
+                sb.toString(),
+                board.showBoard()
```
만약 board.showBoard() 가 두번 이러진다면 이 테스트 케이스는 실패할 것으로 보입니다. 
제가 위에서 언급한 내용입니다.

tmdgusya commented on src/test/java/cooper/chess/BoardTest.java:
```diff
@@ -54,22 +60,10 @@ public void testOutOfBoardSize() {
-        Board board = new Board();
         board.initialize();
-        System.out.println(board.getBoardStatus());
+        System.out.println(board.showBoard());
```
아 이부분에서 showBoard() 함수를 두번 실행시키시고 확인하시면 되겠네요!

tmdgusya reviewed (commented):
쿠퍼! 제가 이번주 마지막 그룹리뷰때 잘집중하지못해서 지금 리뷰하러왔어요 :)

tmdgusya commented on src/test/java/cooper/chess/BoardTest.java:
```diff
@@ -54,22 +60,10 @@ public void testOutOfBoardSize() {
-        Board board = new Board();
         board.initialize();
-        System.out.println(board.getBoardStatus());
+        System.out.println(board.showBoard());
```
쿠퍼 혹시 showBoard() 두번 실행시키면 체스판이 두개 생성되지 않나요?
StringUtils 가 StringBuilder 를 가지고 있는 걸 봤을때 제 예전 코드와 동일한것 같아서, 저도 해당 부분에서 문제를 겪어서.

pbg0205 replied:
아 맞아요, 두 번 출력되요!
생각해보니까 두번 출력되는는 부분 신경쓰이시는게 나중에 게임 진행 도중 위치가 변경할 때도StringBbuiler instance가 존재해서 이전 값 계속 반환할 수도 있겠네요

pbg0205 committed:
style : NEWLINE, NEW_LINE 재명명 및 접근제어자 변경

- 기존의 private 제어 접근자를 public으로 변경하여 줄바꿈에
해당하는 부분에 NEW_LINE을 사용할 수 있도록 하였습니다.

pbg0205 committed:
feat : Piece class equals & hashcode 재정의

- Piece를 값(value)로 관리하는 요구사항에 대해 고민해보았습니다.
값으로 관리한다는 의미는 equals 재정의와 hashcode를 재정의하여 값을
통한 통치 관계를 확인할 수 있도록 구현하는 의미로 받아들여 재정의 작업
을 진행하였습니다.

pbg0205 commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -17,52 +17,70 @@ public Board() {
-            if(color == Color.BLACK) {
-                pieceList.add(Piece.createBlackPawn());
-            }
+        for (int pawnCount = 0; pawnCount < BOARD_SIZE; pawnCount++) {
```
@tmdgusya  로직을 구현할 때 저도 이 부분이 신경쓰였는데 enum을 매개인자를 받으면 생성자를 선언하면 별도의 유효성 검사가 필요없고 팩토리 메서드로 인스턴스를 전달하면 캡슐화되어서 외부의 영향을 받지 않는 부분에서 괜찮겠다 싶어 로직을 구현했어요! 그런데  놓치고 있는 부분이 있을거 같아요; 우선적으로 pieceGroup에 add메서드를 추가해서 로직을 변경하는 것은 저도 동의합니다 감사해요! 혹시 enum으로 매개인자를 받을 때 어떤 유효성 검사를 하면 괜찮을지  생각나는 부분 있으시면 조언 부탁드릴게요!👍

pbg0205 commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -17,52 +17,70 @@ public Board() {
         }
 
-        return pieceList;
+        pieceList.add(Piece.createBlackRook());
```
@tmdgusya  로직을 구현할 때 저도 이 부분이 신경쓰였는데 enum을 매개인자를 받으면 생성자를 선언하면 별도의 유효성 검사가 필요없고 팩토리 메서드로 인스턴스를 전달하면 캡슐화되어서 외부의 영향을 받지 않는 부분에서 괜찮겠다 싶어 로직을 구현했어요! 그런데  놓치고 있는 부분이 있을거 같아요; 우선적으로 pieceGroup에 add메서드를 추가해서 로직을 변경하는 것은 저도 동의합니다 감사해요! 혹시 enum으로 매개인자를 받을 때 어떤 유효성 검사를 하면 괜찮을지  생각나는 부분 있으시면 조언 부탁드릴게요!👍

pbg0205 replied:
@tmdgusya  로직을 구현할 때 저도 이 부분이 신경쓰였는데 enum을 매개인자를 받으면 생성자를 선언하면 별도의 유효성 검사가 필요없고 팩토리 메서드로 인스턴스를 전달하면 캡슐화되어서 외부의 영향을 받지 않는 부분에서 괜찮겠다 싶어 로직을 구현했어요! 그런데  놓치고 있는 부분이 있을거 같아요; 우선적으로 pieceGroup에 add메서드를 추가해서 로직을 변경하는 것은 저도 동의합니다 감사해요! 혹시 enum으로 매개인자를 받을 때 어떤 유효성 검사를 하면 괜찮을지  생각나는 부분 있으시면 조언 부탁드릴게요!👍

kyupid commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -1,111 +1,119 @@
 
-    private Map<Color, List<Pawn>> createPawnGroup() {
-        Map<Color, List<Pawn>> pawnListMap = new HashMap<>();
+    private Map<Color, List<Piece>> createPawnGroup() {
+        Map<Color, List<Piece>> pawnListMap = new HashMap<>();
```
전체적으로 전부 Map 을 사용하셨네요!

pbg0205 replied:
색상에 따라서 구분하고 싶어서 HashMap으로 작성해봤어요! 아 그리고 이전 피드백에서 해당 자료구조가 복잡해질수도 있다는 피드백을 받아서 별도로 Wrapper클래스로 분리해서 관리하고 있어요. Piece class 내부에서 계속 관리하다보니까 내부 로직이 너무 복잡해지더라구요; 원래는 일급컬렉션 피드백을 받아서 작성했지만 정확히 일급컬렉션이다!!!! 라는 확신이 아직까지도 없네요ㅠ

kyupid commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -1,111 +1,119 @@
 
-        if(pawnList.size() == 0) {
+        if (pieceList.size() == 0) {
             throw new IllegalArgumentException("size가 0입니다.");
         }
 
-        if ((0 > index) || (index >= pawnList.size())) {
+        if ((0 > index) || (index >= pieceList.size())) {
             throw new IllegalArgumentException("범위를 벗어나는 인덱스 입니다.");
```
우와 예외까지 작성하셨군요 👏

pbg0205 replied:
어?! 지금 pieceList에 모든 기물이 다 담겨야해서 사이즈가 16이 넘어가면 안되는데 이 부분 수정 못했네요ㅠㅋㅋㅋ
kyu 덕분에 잘못된 부분 찾았어요ㅋㅋㅋ 감사합니다!

pbg0205 replied:
아 저 부분 add부분으로 착각했네요ㅋㅋㅋ 그래도 kyu가 리뷰 달아줘서 다시 한번 제 코드 작성한 이유에 대해서 다시 한번 생각했어요👍

kyupid commented on src/main/java/cooper/chess/utils/StringUtils.java:
```diff
@@ -0,0 +1,15 @@
+package cooper.chess.utils;
+
+public class StringUtils {
+    public static final String NEW_LINE = System.getProperty("line.separator");
+    private static final StringBuilder stringBuilder = new StringBuilder();
```
아하 StringBuilder 가 유틸안에 같이 있었구나

pbg0205 replied:
아 근데 저 부분은 좀 제가 만져봐야 될 것 같아요,,, 매 턴이 지날때 마다 새로운 상태를 출력하는 것을 생각해볼 때 그렇게 효율적인 방법은 아닌거 같아요;

honux77 replied:
넵 맞아요. ㅎㅎ 유틸 클래스에 상태가 변할 수 있는 객체가 있는 것은 좋은 생각이 아닌 것 같아요.

pbg0205 replied:
넵ㅠ 그래서 StringUtils에 StringBuilder을 Board 클래스 showBoard 내에 지역변수로 변경하여서 각 턴마다 상태를 출력할 수 있도록 로직 변경했습니다!

kyupid commented on src/main/java/cooper/chess/piece/Color.java:
```diff
@@ -1,32 +1,6 @@
-        return colorName;
-    }
-
-    public static Color of(String colorName) {
```
메서드 이름이랑 매개변수 사이에 `of`가 혹시 뭔지 알수있을까요?

pbg0205 replied:
of 메서드는 Enum 내부에 있는 멤버(변수)를 통해서 Enum 상수를 찾는 메서드에요!
Enum 내부에 valueOf라는 메서드가 있는데 이거는 문자열 값이 일치하는 상수를 반환하는 메서드인데 저는 상수 내부에 있는 변수로 상수를 찾고 싶어서 별도로 정적 메서드로 탐색하는 메서드로 작성했었어요ㅋㅋ
그런데 생성자를 둘다 enum으로 받아서 별로 필요없겠다 싶어서 제거해버렸어요!

kyupid commented on src/main/java/cooper/chess/piece/Piece.java:
```diff
@@ -0,0 +1,84 @@
+
+import java.util.Objects;
+
+public class Piece {
```
우왕 되게 깔끔하게 작성하셨네요. 컬러따로 타입따로 이넘으로 빼주니까 이렇게 깔끔하게 작성할수있는건가? 내껀 엄청 지저분한데 ㅠㅋㅋㅋ

pbg0205 replied:
처음에 저도 문자열로 받아서 처리하는 방식으로 진행했는데 enum으로 빼는게 더욱 효율적이라고 생각했어요! 색상하고 기물은 값이 불연속적이고 갯수가 정해져 있는 값이라 enum으로 작성하는 것이 직관적으로 좋아보이고 정확한 값을 전달할 수 있어서 변경했어요
유효성 검사에서 별도 안해줘도 될거 같은데 이 부분은 제가 아직 고려하지 못하는 부분이 있을거 같아요ㅠ

kyupid commented on src/main/java/cooper/chess/piece/Piece.java:
```diff
@@ -0,0 +1,84 @@
+        return type.getRepresentaion(this.color);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o instanceof Piece){
+            return false;
+        }
+        Piece piece = (Piece) o;
+        return color == piece.color && type == piece.type;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(color, type);
+    }
+}
```
와 뭔가 오버라이딩하셨네요. 어디에 쓰일지 궁금합니다. 계속 스크롤내리는중.

pbg0205 replied:
equals재정의를 하면 값을 통해서 객체가 같다는 걸로 인지하게 할 수 있어요! 아이폰 두개가 있는데 컴퓨터는 시리얼 넘버로 같다고 한다면 저희는 그냥 기종보고 같은 폰인걸로 인지하잖아요. 그런거 처럼 사람이 사물을 볼 때 같은 기준을 설정하는 거라고 생각하시면 될 거 같아요! 요구사항에 객체를 값으로 비교한다는 내용을 보고 작성한 내용이에요!
 hashcode 재정의는 Hash를 사용하는 자료구조에서는 hash 값을 통해서 값이 같은지를 비교하는거라 별도로 hashcode 재정의를 해주지 않으면 hash와 관련된 Collection에서 다른 값으로 인지해버려서 재정의를 해줘야 해요!

kyupid commented on src/main/java/cooper/chess/piece/Type.java:
```diff
@@ -0,0 +1,21 @@
+    QUEEN('q'),
+    KING('k');
+
+    private final Character representation;
```
와 char 로 선언안하시고 Character 로 선언하신 이유가 따로 있으신가요?

pbg0205 replied:
Charater 사용한 이유는 toUpperCase나 toLowerCase() 메서드를 사용하려고 wrapper 클래스로 선언해줬어요. primitive type으로 선언해서 사용하면 ascii code 차이만큼 (대소문자 차이가 26이었나,,,?) 해서 값을 빼고 더해줘야 되는데 메서드로 사용하는 것이 가독성도 좋고 사용하기도 편리해서 wrapper class를 사용했어요!

honux77 replied:
오토박싱이 되니까 char 사용해도 상관없긴 한데요 ㅎㅎ.

kyupid commented on src/main/java/cooper/chess/piece/Type.java:
```diff
@@ -0,0 +1,21 @@
+
+    public char getRepresentaion(Color color) {
+        return color == Color.WHITE ?
+                Character.toUpperCase(representation) : representation ;
```
Character 참조형으로 쓰면 UpperCase를 쓸수있어서 그런거군요??맞낭?? 오호 이런식으로 사용하면 뭔가 제 코드도 간결화 할 수 있을거 같은 좋은느낌?

kyupid reviewed (commented):
안녕하세요 ㅎㅎ
스크롤하면서 그냥 캐쥬얼하게 짧막하게 감탄사들과 간단한 질문남겼어요 ㅎㅎ 피드백다운 피드백이 없어서 도움은 안되겠지만요..ㅠ 그냥 쿠퍼 코드도 보고싶었습니당

honux77 commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -1,111 +1,119 @@
 public class Board {
-    private static final int PAWN_MAX_SIZE = 8;
+    private static final int BOARD_SIZE = 8;
+    private static final String BLANK = "........";
```
BLANK는 static 변수보다 지역변수가 더 어울리지 않을까요? 가능하면 static 변수는 적은 편이 좋아요.

pbg0205 replied:
해당 변수 showBoard 메서드로 이동했습니다! 코멘트 감사합니다!

honux77 commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -1,111 +1,119 @@
 
-    private Map<Color, List<Pawn>> createPawnGroup() {
-        Map<Color, List<Pawn>> pawnListMap = new HashMap<>();
+    private Map<Color, List<Piece>> createPawnGroup() {
+        Map<Color, List<Piece>> pawnListMap = new HashMap<>();
```
1. PieceGroup 이라는 클래스 이름이 살짝 어색해요.  근데 저도 더 좋은 이름은 못 찾겠네요.
2. 보드에서 맵을 생성해서 넘겨주는데 그 부분도 상당히 어색해요. 불필요한 의존성도 발생하는데 그룹이 자기자신을 초기화하는게 더 나아 보입니다.

pbg0205 replied:
PieceGroup은 아직 좋은 이름이 떠오르지 않습니다,,, 이후 미션에서 더 좋은 이름이 떠오르면 변경해보도록 해보겠습니다.
그리고 맵을 넘겨주는 부분이 미션3을 구현하면서 제일 고민했던 부분이었습니다. wraper 클래스로 만드는 피드백과 함께 레퍼런스로 일급 컬렉션을 제공 받았는데 기준이 명확하지 않아 일급컬렉션 어설프게 따라해서 발생한 부분인 것 같습니다. 리뷰받고 PieceGroup 내부에서 모든 기물을 초기화할 수 있도록 로직 변경하였습니다. 리뷰 감사합니다!

honux77 commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -1,111 +1,119 @@
-    private List<Pawn> initPawnList(Color color) {
-        List<Pawn> pawnList = pawnGroup.getPawnList(color);
+    private void initWhitePiece() {
+        List<Piece> pieceList = pieceGroup.getPawnList(Color.WHITE);
```
이런 부분도 상당히 어색한데요. PieceGroup에서 굳이 리스트를 받아와서 추가하는 것보다 pieceGroup에서 직접 add하는 편이 자연스러울 것 같습니다.
`pieceGroup.add(Piece.createWhiteRook());`

honux77 commented on src/main/java/cooper/chess/piece/Piece.java:
```diff
@@ -0,0 +1,84 @@
+        this.type = type;
+    }
+
+    public static Piece createWhitePawn() {
+        return new Piece(Color.WHITE, Type.PAWN);
+    }
+
+    public static Piece createBlackPawn() {
+        return new Piece(Color.BLACK, Type.PAWN);
+    }
```
이런식으로 하는 것도 상당히 괜찮은 것 같아요. 소스코드가 길어져서 타이핑할때는 인내심이 필요하겠지만요 ㅎㅎ. 
(저는 하다가 포기했습니다.) 👍

pbg0205 replied:
타이핑할 때 조금 힘들었지만 복붙하다보니 한계가 오기 전에 간신히 완료했습니다ㅋㅋ👍

honux77 commented on src/main/java/cooper/chess/piece/PieceGroup.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.List;
+import java.util.Map;
+
+public class PieceGroup {
```
기껏 생성해 놓은 PieceGroup의 역할이 상당히 모호하고, Baord에는 여전히 List가 존재함으로 코드의 복잡도만 증가시켰어요. 가능하다면 보드에는 List와 관련된 코드를 최소화하고 책임을 PeiceGroup쪽으로 넘기면 좋을 것 같습니다.

pbg0205 replied:
네 말씀해주신대로 PieceGroup으로 책임 변경하였습니다! 코멘트 감사합니다!

honux77 reviewed (changes requested):
전반적으로 코드가 깔끔하고 좋은데 PieceGroup이 뭔가 완성도를 크게 떨어뜨린 느낌입니다.
관련 부분 코드를 먼저 개선하면 좋겠어요!

pbg0205 committed:
refactor : StringUtils 줄바꿈 메서드 변경

- 기존에 StringUtils 내부에 존재하던 StringBuilder 정적 변수를 제거
하였습니다.

- 줄바꿈 메서드를 담당하는 appendNewLine 메서드 내부의 로직을 문자열을
받아 줄바꿈을 하도록 로직을 변경하였습니다.

pbg0205 committed:
refactor : PieceGroup으로 메서드 및 초기화 이동

- Board에서 매개인자로 받아 PieceGroup을 초기화 하던 방법을
PieceGroup에서 모든 기물을 초기화하는 방법으로 변경하였습니다.

- 기존에 모든 기물을 초기화하던 initialize 메서드를 제거하였습니다.

pbg0205 committed:
test : PieceGroup 초기화에 따른 테스트 코드 변경

- PieceGroup에서 초기화하므로 사이즈가 0일 때, 발생하는 예외 테스트
코드는 이번 미션 단계에서 해당 사항이 아니라고 생각해 테스트 코드를
제거하였습니다.

- PieceGroup에 따른 흰색과 검은색의 예외 코드를 추가 했습니다.

jihye-woo commented on src/main/java/cooper/chess/piece/PieceGroup.java:
```diff
@@ -0,0 +1,89 @@
+        initialize();
+    }
+
+    private void initialize() {
+        initWhitePiece();
+        initBlackPiece();
+    }
```
색깔별로 초기화 구간을 나누니까 더 깔끔하고 보기 좋네요! 저는 그냥 열거해놓았는데ㅋㅋ큐ㅠㅠ
그리고 물론 미래의 일은 아직 생각하지 않아도 되지만..! initWhitePiece랑 initBlackPiece 로직이 비슷해서 나중에 좀 더 간결하게 만들기도 좋을 것 같아요!
좋은 팁 얻어갑니다! 감사합니다😀

honux77 replied:
"길어지면 나누자" 가 대체로 좋은 생각인 것 같습니다.

jihye-woo replied:
이 부분은 확실히 나누는게 맞는 것 같아서 바아로 수정했습니다!😀

honux77 reviewed (approved):
제가 보기엔 상당히 깔끔해진 것 같아요. 다만 이제 보드의 책임이 무엇인지에 대해서 또 생각해 봐야하겠네요 ㅎㅎ 
수정한 내용 다음단계에도 반영해 주시고 추가로 개선할 내용 개선하면 좋을 것 같아요!
